src/uu/join/src/join.rs

```diff
@@ -334,28 +334,34 @@ impl<'a> State<'a> {
         key: usize,
         line_ending: LineEnding,
         print_unpaired: bool,
-    ) -> State<'a> {
-        let f = if name == "-" {
+    ) -> Result<State<'a>, JoinError> {
+        let file_buf = if name == "-" {
             Box::new(stdin.lock()) as Box<dyn BufRead>
         } else {
             match File::open(name) {
                 Ok(file) => Box::new(BufReader::new(file)) as Box<dyn BufRead>,
-                Err(err) => crash!(1, "{}: {}", name.maybe_quote(), err),
+                Err(err) => {
+                    return Err(JoinError::UnorderedInput(format!(
```

Comment:
`UnorderedInput` is probably not quite right is it? It is the most convenient solution and it works, but it doesn't make sense :smile:

`UError` has a way of adding some context to an io error like so:

```rs
let file = File::open(name).map_err_context(|| name.maybe_quote())?;
```

Then, the function should return a `UResult` (and all the callers of this function as well). I think that should be fairly easy to do.

That is how other utils usually do this. Could you give that a try?

Reply:
Absolutely, thanks for the feedback!